pkg/bootstrap/bootstrap.go

```diff
@@ -29,79 +29,75 @@ const (
-	nodes map[node.ID]*NodeStatus[T]
-	// newBootstrapMsg is a factory function that returns a new bootstrap message
-	newBootstrapMsg NewBootstrapMessageFn
-	resendInterval  time.Duration
+	// newBootstrapRequest returns a new bootstrap message
+	newBootstrapRequest NewBootstrapRequestFn
+	resendInterval      time.Duration
 
 	// For test only
 	currentTime func() time.Time
 }
 
 // NewBootstrapper create a new bootstrapper for a distributed instance.
-func NewBootstrapper[T any](id string, newBootstrapMsg NewBootstrapMessageFn) *Bootstrapper[T] {
+func NewBootstrapper[T any](id string, newBootstrapMsg NewBootstrapRequestFn) *Bootstrapper[T] {
 	return &Bootstrapper[T]{
-		id:              id,
-		nodes:           make(map[node.ID]*NodeStatus[T]),
-		bootstrapped:    false,
-		newBootstrapMsg: newBootstrapMsg,
-		currentTime:     time.Now,
-		resendInterval:  defaultResendInterval,
+		id:                  id,
+		nodes:               make(map[node.ID]*node.Status[T]),
+		bootstrapped:        false,
+		newBootstrapRequest: newBootstrapMsg,
+		currentTime:         time.Now,
+		resendInterval:      defaultResendInterval,
 	}
 }
 
-// HandleNewNodes add node to bootstrapper and return messages that need to be sent to remote node
-func (b *Bootstrapper[T]) HandleNewNodes(newNodes []*node.Info) []*messaging.TargetMessage {
-	msgs := make([]*messaging.TargetMessage, 0, len(newNodes))
+// HandleNewNodes updates the bootstrapper with the current set of active nodes.
+// It returns the IDs of newly added nodes, removed nodes, messages to be sent, and any cached bootstrap responses.
+func (b *Bootstrapper[T]) HandleNewNodes(activeNodes map[node.ID]*node.Info) (
+	addedNodes []node.ID,
+	removedNodes []node.ID,
+	messages []*messaging.TargetMessage,
+	responses map[node.ID]*T,
+) {
 	b.mutex.Lock()
 	defer b.mutex.Unlock()
-	for _, info := range newNodes {
-		if _, ok := b.nodes[info.ID]; ok {
+	for id, info := range activeNodes {
+		if _, ok := b.nodes[id]; ok {
 			continue
 		}
 		// A new node is found, send a bootstrap message to it.
-		b.nodes[info.ID] = NewNodeStatus[T](info)
-		log.Info("maintainer found a new node",
-			zap.String("changefeed", b.id),
-			zap.String("nodeAddr", info.AdvertiseAddr),
-			zap.Any("nodeID", info.ID))
-		msgs = append(msgs, b.newBootstrapMsg(info.ID))
-		b.nodes[info.ID].lastBootstrapTime = b.currentTime()
+		b.nodes[id] = node.NewStatus[T](info)
+		messages = append(messages, b.newBootstrapRequest(id, info.AdvertiseAddr))
+		b.nodes[id].SetLastBootstrapTime(b.currentTime())
+		addedNodes = append(addedNodes, id)
 	}
-	return msgs
-}
 
-// HandleRemoveNodes remove node from bootstrapper,
-// finished bootstrap if all node are initialized after these node removed
-// return cached bootstrap
-func (b *Bootstrapper[T]) HandleRemoveNodes(nodeIDs []node.ID) map[node.ID]*T {
-	b.mutex.Lock()
-	for _, id := range nodeIDs {
-		status, ok := b.nodes[id]
-		if ok {
-			delete(b.nodes, id)
+	if len(addedNodes) > 0 {
```

Comment:
So if a new node joins, the bootstrapped will be false until maintainer get the response from the new node? 
It's seems different with the previous implementation. Will this introduce correctness issues?